ksql-engine/src/main/java/io/confluent/ksql/physical/KsqlQueryBuilder.java

```diff
@@ -0,0 +1,143 @@
+import org.apache.kafka.common.serialization.Serde;
+import org.apache.kafka.connect.data.Schema;
+import org.apache.kafka.streams.StreamsBuilder;
+
+public final class KsqlQueryBuilder {
+
+  private final StreamsBuilder streamsBuilder;
+  private final KsqlConfig ksqlConfig;
+  private final ServiceContext serviceContext;
+  private final ProcessingLogContext processingLogContext;
+  private final FunctionRegistry functionRegistry;
+  private final QueryId queryId;
+  private final LinkedHashMap<String, Schema> schemas = new LinkedHashMap<>();
+
+  public static KsqlQueryBuilder of(
+      final StreamsBuilder streamsBuilder,
+      final KsqlConfig ksqlConfig,
+      final ServiceContext serviceContext,
+      final ProcessingLogContext processingLogContext,
+      final FunctionRegistry functionRegistry,
+      final QueryId queryId
+  ) {
+    return new KsqlQueryBuilder(
+        streamsBuilder,
+        ksqlConfig,
+        serviceContext,
+        processingLogContext,
+        functionRegistry,
+        queryId
+    );
+  }
+
+  private KsqlQueryBuilder(
+      final StreamsBuilder streamsBuilder,
+      final KsqlConfig ksqlConfig,
+      final ServiceContext serviceContext,
+      final ProcessingLogContext processingLogContext,
+      final FunctionRegistry functionRegistry,
+      final QueryId queryId
+  ) {
+    this.streamsBuilder = requireNonNull(streamsBuilder, "streamsBuilder");
+    this.ksqlConfig = requireNonNull(ksqlConfig, "ksqlConfig");
+    this.serviceContext = requireNonNull(serviceContext, "serviceContext");
+    this.processingLogContext = requireNonNull(processingLogContext, "processingLogContext");
+    this.functionRegistry = requireNonNull(functionRegistry, "functionRegistry");
+    this.queryId = requireNonNull(queryId, "queryId");
+  }
+
+  public ProcessingLogContext getProcessingLogContext() {
+    return processingLogContext;
+  }
+
+  public ServiceContext getServiceContext() {
+    return serviceContext;
+  }
+
+  public KsqlConfig getKsqlConfig() {
+    return ksqlConfig;
+  }
+
+  public FunctionRegistry getFunctionRegistry() {
+    return functionRegistry;
+  }
+
+  public StreamsBuilder getStreamsBuilder() {
+    return streamsBuilder;
+  }
+
+  public QuerySchemas getSchemas() {
+    return QuerySchemas.of(schemas);
+  }
+
+  public KsqlQueryBuilder withKsqlConfig(final KsqlConfig newConfig) {
+    return of(
+        streamsBuilder,
+        newConfig,
+        serviceContext,
+        processingLogContext,
+        functionRegistry,
+        queryId
+    );
+  }
+
+  public QueryContext.Stacker buildNodeContext(final PlanNodeId id) {
+    return new QueryContext.Stacker(queryId)
+        .push(id.toString());
+  }
+
+  public Serde<GenericRow> buildGenericRowSerde(
+      final KsqlTopicSerDe topicSerDe,
+      final Schema schema,
+      final QueryContext queryContext
+  ) {
+    final String loggerNamePrefix = QueryLoggerUtil.queryLoggerName(queryContext);
+
+    track(loggerNamePrefix, schema);
```

Comment:
This is sneaky - by looking at a method `buildGenericRowSerde` I wouldn't expect side-effects (tracking) if called multiple times. Can we just expose `track(QueryContext, Schema)` as a separate method from building the serde?

Reply:
Not really, the point is that we want to track all built SerDe classes. Hence we track it in the build method. 

Also remember, in affect this is test-only code in Production. (Not something I like at all). To minimizing its presence to within this one class is a good thing. IMHO.